### ISSUE TITLE
test(model): localize strict Metal P4 layer3 residual and attention divergence

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12927]
Parent: #12349
Coordinates with: #12853
Coordinates with: #12265
Coordinates with: #12859
Coordinates with: #12867

## Parent context
#12349 requires exact native P4 verification.

## Current state
#12920 physical native M3 Pro run passed six focused/P1 checks; H=D=S for layer0 mixed QKV and all codes/scales matched. Full strict model parity remains RED: layer3 KRoPE[4101] is row0 channel5; KRaw/V[4224] is row1 channel0, with prefix32 and KV width128. A combined hidden-history assertion does not distinguish missing rows from token/value differences. Foundation is #12920 full manifest ab896f4bd0b5051b607da2d7f578cef57fd9b5c7555e53af863e6f2be8fdccf8, q8 d4794900b23c817ecb2b251328a48d3ddf1a2e6ee930b36606cd6d3323e4dd9a, q4k f7c3bdc617c66dda92acf0e5614ff0495502c5c6edff23520396eff3354a54df, strict test d682fe65cde687886a0368f3c6bea8eed2e6b5d03343b3b7f165212010742b13.

## For / Problem / Today / Better because
For native Metal MTP users, identify the first differing layer3 arithmetic boundary before changing kernels. Today final logits and caches cannot separate incoming residual, normalization, projection and RoPE drift. Better because actual graph values and native serial intermediates are compared in causal order with exact bits and independently bound to the real serial cache.

## Why this is next
Earlier codes/scales/layer0 projection differences have direct physical repair evidence. Layer3 incoming residual remains unobserved; no Q4/Q6 cause is presumed.

## Working spine
Two fresh diagnostic sessions with the same prefix32 and four tokens3,5,7,11, isolated from strict target/serial. A test-local source-matched P4 graph executes layers0–2, then layer3 norm, packed Q/K/V and FullAttention, stopping BEFORE layer3 AddInPlace mutates the incoming residual. One final packed read captures actual residual, norm, packedQ, projectedK, projectedV, KRaw and KPost. FullAttention.V aliases projectedV and is read once. A native serial replay uses the unchanged concrete sessionQ4KKernel and existing blockStep layers0–2, native mixed-QKV/projection routes, and existing norm/bias/QKNorm/RoPE semantics. Copy intermediates before mutation. Bind replay KRaw/KPost/V to actual original serial.Step cache rows. Compare all four rows by bits, hashes, mismatch counts and first row/token/channel. Pure host RoPE replay of captured graph KRaw provides a separate arithmetic comparison. Report hidden token lengths, row counts, first token mismatch and overlapping row values separately.

## Core through-line
Observe actual native operands and preserve the original exact acceptance oracle. No production arithmetic, dispatch or tolerance changes.

## Scope
```routing
lane: model
paths: ["internal/model/metal_mtp_verify_panel_darwin_test.go"]
expected_steps: 3
```

## Likely files
- internal/model/metal_mtp_verify_panel_darwin_test.go

## Scope class
Work unit: leaf. One existing test file, bounded test-local diagnostic; no generic numerical tracer.

## Definition of done
- [ ] Managed one-file candidate independently reviewed against source-matched production sequence and native serial replay.
- [ ] Frozen diff and complete source manifest before separate HIL grant.
- [ ] One physical invocation captures all stage comparisons and graph accounting without weakening strict acceptance.
- [ ] Independent result reconciliation to parent; no presumption of full parity.

## Gold-plating boundary
No production edits, config mutation, CPU engine fallback, kernel fix, broad tracer, tolerance change or performance claim. Do not confuse later-mutated graph x with its layer3 input. Diagnostic sessions never mutate strict target or serial state; GDN checkpoints/backups close on all paths. Restore any global profiler state.

## Acceptance gate
Require one completed graph command, one FinishRead of seven unique buffers, exactly24576 read bytes, no intermediate waits/readbacks, and separate accounting from production. Source audit must bind mirrored layers0–2 and layer3 norm/QKV/attention sequence to encodeQwen35MetalMTPP4. Serial replay must bind its cache values to original actual serial.Step caches, with all intermediate copies made before mutation. All four rows and real weight kinds/config flags logged. Known downstream strict failure remains failure; diagnostics classify evidence separately.

## Done condition
One independently reviewed diagnostic source and bounded physical receipt localizes the first observed differing stage. Parent #12349 remains open until its original strict gates pass.

## Witness
Proposed, NOT authorized for execution:
env -u FAK_METAL_MPS CGO_ENABLED=1 GOMAXPROCS=2 GOFLAGS=-p=1 GOCACHE=/tmp/fak-12853-reduction-run/cache GOTMPDIR=/tmp/fak-12853-reduction-run/tmp go test ./internal/model -run '^TestMetalMTPP4StrictSerialParity$' -count=1 -timeout=60s -json

## Budget
One invocation, native60 seconds; owned process group TERM230/KILL240/census270/report300. Added work: two32-token session preparations, one truncated P4 graph, twelve native P1 blocks, four native QKV calls, one24576-byte final packed read (4096 residual +4096 norm +8192 packedQ +2048 projectedK +2048 projectedV +2048 KRaw +2048 KPost). No extra models/benchmarks/retry/device campaign. Preserve resident80431, fresh owner/source/lease/resources required. Resource/device/timeout failures are inconclusive. No timing claims. Independent proposal review authorizes source authoring only; applied candidate and budget require separate final review/grant before HIL.

## Duplicate search
REST search across open/closed issues for layer3 and layer 3 diagnostic found no exact leaf. #12853 is the existing layer0 normalization diagnostic, #12265 is a general numerical tracer with broader contracts. This child is a bounded fixture-specific layer3 experiment, not implementation of that tracer.

## Problem frame
Centrality: Core
Portfolio tier: 2 serving
P1: preserved - preserve exact native state equality.
P2: preserved - no performance claim.
P3: preserved - one test file and bounded experiment.
P4: advanced - witness actual incoming graph values and native route.

## Work estimate
Estimate: 1 point.

## Overall completion contribution
Contribution: 1/4 points within #12349; planning only.

## Completion standard
experiment

## Closure binding
A witnessed explicit-path commit cites this child and (fak model), or independently record a non-landing experiment disposition. Full parent closure requires its original strict gates.

**Comment by anthony-chaudhary** [comment]
Source-only candidate authored after independent proposal review. No compilation, tests, or HIL have run for this child.

- Exactly one authored file: `internal/model/metal_mtp_verify_panel_darwin_test.go`, 346 additions / 5 removals. Applied delta SHA256 `1ed51da81b448d39a4fa60b7a0a5bfcde67cd01c5d5281e739bcc506583aa7d2`; test SHA256 `b100998489ce7836a19ef739f3719101703bf6b470da6aac7847b2603f1fd83a`.
- Complete 1702-file source manifest SHA256 `f6ed5e7a263a281dd3f6abc916f886a741295a59f7ba4e9842cabcd4eb53b36b`. All other1701 foundation files unchanged from #12920, including Q8 d4794900 and Q4_K f7c3bdc6.
- Graph mirrors production layers0–2 and layer3 norm/projection/attention, stops before layer3 residual mutation, and reads seven unique buffers totaling24576 bytes once. FullAttention V alias is checked and read once. Diagnostic accounting is separate.
- Two fresh diagnostic sessions own replay/graph state. Graph, checkpoints and sessions close in dependency order; only the diagnostic session profiler changes and is restored. Serial intermediates are copied before mutation, with KRaw/KPost/V bound to actual serial.Step cache rows32–35.
- The diagnostic is placed after all original strict continuation assertions, then excludes continuation row36 from cache binding. Source comparison confirms the original strict comparator and transaction/serial/continuation assertions remain verbatim. Hidden-history checks now distinguish lengths, token mismatch and overlapping row values; nil-row counts and token suffixes are logged.
- `gofmt -l` empty and `git diff --check` passed. Managed one-file lease released. Applied source requires independent review, then a separate explicit bounded HIL grant. Proposed native60/TERM230/KILL240/census270/report300 budget unchanged; no automatic retry or performance claim.

Next check: independent applied-source review against production graph and serial dispatch before scheduling one physical witness. Parent #12349 remains open.

**Comment by anthony-chaudhary** [comment]
The single authorized #12927 attempt ended INCONCLUSIVE before HIL: executor99016 / owned PID+PGID39836 exited1 after154.173s. Go compilation failed writing model `$WORK/b123/_pkg_.a` and ctxmmu `$WORK/b200/_pkg_.a`: `no space left on device`. There were zero named test run events and no new arithmetic or hardware result.

The coordinator requested a capacity abort as shared disk reached226MiB free. The immediate owned-identity check found the process already terminal, so no signals were sent. Final census found no owned group or tracked descendants. Exact one-file lease released; all1702 frozen source files unchanged (manifest f6ed5e7a, patch1ed51da8, testb1009984). No retry, peer control, source deletion or cache deletion occurred. The exact run GOTMPDIR was subsequently empty; no leftover cleanup was needed.

Captured receipt SHA256 `9170ad52c586dd1df5d1b51fd8c8fb6c1dc18707abb66425efcdd03221236fc7`; raw JSONL `89da20b01fbc8a34c5baaf94ce8cbc24169421673f847eebd877b1384fdd8902`; supervision `04d64ceb0e27d3aea21def71ecf37541b8bf7188c641295897c58ff25ecfee81`. Independent foreign test/server activity was explicitly allowed for functional-only execution and was preserved. No performance claim.

Next: recover disk capacity through separately owned work, then obtain a new explicit bounded attempt grant for the unchanged reviewed candidate. #12927 and parent #12349 remain open; this attempt adds no parity evidence.

**Comment by anthony-chaudhary** [comment]
Second separately authorized attempt completed the physical diagnostic on Apple M3 Pro, engine fak-native. Executor59207 / PID+PGID56795 exited1 after74.571s: one strict test ran and failed, zero skips. No resource/device/timeout failure, capacity abort or signal. The first ENOSPC attempt remains preserved separately.

Diagnostic validity: all four replay rows match original serial.Step KRaw/KPost/V bits (12 comparisons of128 elements). Graph captured its actual incoming layer3 residual before mutation, with one command/wait/packed read of seven unique buffers totaling24576 bytes;64864 upload bytes,60 encoders,zero intermediate waits/readbacks; V alias counted once. Serial route recorded48 native events,zero fallback,12 resident GDN acceptances; CPU orchestration retained the concrete sessionQ4KKernel. QKNorm=false; Q/K weights areQ8 andV isQ4_K.

Mismatch counts by row (tokens3,5,7,11):

| Boundary | Row0 | Row1 | Row2 | Row3 |
|---|---:|---:|---:|---:|
| Incoming residual, width256 | 0 | 0 | 0 | 213 |
| Input norm, width256 | 0 | 235 | 0 | 228 |
| PackedQ, width512 | 0 | 380 | 0 | 355 |
| KRaw, width128 | 0 | 97 | 0 | 97 |
| V, width128 | 0 | 109 | 0 | 115 |
| KPost, width128 | 31 | 105 | 32 | 100 |
| Host RoPE on graph KRaw versus graph KPost | 31 | 33 | 32 | 30 |

Causal scope: row3 already differs before layer3, requiring earlier-layer localization. Row1 has identical incoming input but differing norm (firstchannel0 serialbf7f9a44 versus graphbf7f9a46), directly identifying a norm boundary difference. Rows0/2 remain exact throughKRaw but differ atKPost; replaying host RoPE on the actual graphKRaw differs on all rows, separating postKRaw arithmetic from upstream errors. These observations do not establish a Q4/Q6 projection defect or select a production fix.

The original strict logits rows1/3, committed boundary and layer3 caches still fail. Layer0 H=D=S remains exactcf28fa88 and all1024 codes/all32scale bits match. Hidden bookkeeping is now explicit: serial0rows/0tokens; target36rows/36tokens,32nilprefix rows and suffix[3,5,7,11] aftercommit/continuation. This is not evidence of losing the verified suffix.

Receipt SHA256 `00fe57e8d3c82c1a86fc5e660b9bbcd18499627eb683952aeeb1b1354881c627`; rawJSONL `112ecbb29471d34f378ea86615a9b9cbea9e58cc08432d5a227735d0755ee29c`; supervision `5e38392d950bd01d7ef6bb78afdb5022830c248b3506331ecdaef31a1eb76a08`. All1702 source files unchanged (manifestf6ed5e7a/testb1009984/patch1ed51da8). Fourteen capacity samples stayed above2GiB (minimum6.86GiB). Owned group/descendants absent; exact lease released. No timing/performance claim amid permitted peer overlap.

Next: independently validate causal interpretation and scope the smallest production arithmetic experiment. No further run or source edit is authorized yet; parent #12349 remains open.
